File: base/src/main/java/de/whitefrog/frogr/persistence/FieldDescriptor.java
package de.whitefrog.frogr.persistence;

import de.whitefrog.frogr.helper.ReflectionUtil;
import de.whitefrog.frogr.model.Base;
import de.whitefrog.frogr.model.Model;
import de.whitefrog.frogr.model.annotation.*;
import de.whitefrog.frogr.model.relationship.Relationship;

import java.lang.reflect.Field;
import java.util.Collection;

public class FieldDescriptor<T extends Base> {
  private AnnotationDescriptor annotations;
  private Field field;
  private boolean collection;
  private boolean model;
  private boolean relationship;
  private Class<T> baseClass;

  public FieldDescriptor(Field field) {
    field.setAccessible(true);
    this.field = field;
    this.collection = Collection.class.isAssignableFrom(field.getType());
    
    AnnotationDescriptor descriptor = new AnnotationDescriptor();
    descriptor.indexed = field.getAnnotation(Indexed.class);
    descriptor.notPersistent = field.isAnnotationPresent(NotPersistent.class);
    descriptor.relatedTo = field.getAnnotation(RelatedTo.class);
    descriptor.unique = field.isAnnotationPresent(Unique.class);
    descriptor.fetch = field.isAnnotationPresent(Fetch.class);
    descriptor.required = field.isAnnotationPresent(Required.class);
    descriptor.nullRemove = field.isAnnotationPresent(NullRemove.class);
    descriptor.blob = field.isAnnotationPresent(Blob.class);
    descriptor.uuid = field.isAnnotationPresent(Uuid.class);
    descriptor.lazy = field.isAnnotationPresent(Lazy.class);
    descriptor.relationshipCount = field.getAnnotation(RelationshipCount.class);

    this.annotations = descriptor;
    
    if(this.collection) {
      this.baseClass = (Class<T>) ReflectionUtil.getGenericClass(field);
    } else {
      this.baseClass = (Class<T>) field.getType();
    }
    
    this.model = Model.class.isAssignableFrom(baseClass);
    this.relationship = Relationship.class.isAssignableFrom(baseClass);
  }
  
  public AnnotationDescriptor annotations() {
    return annotations;
  }
  public boolean isCollection() {
    return collection;
  }
  public Class<T> baseClass() {
    return baseClass;
  }
  public Field field() {
    return field;
  }
  public String getName() {
    return field.getName();
  }

  public boolean isModel() {
    return Model.class.isAssignableFrom(baseClass);
  }

  public boolean isRelationship() {
    return relationship;
  }

  @Override
  public String toString() {
    return "Field: \"" + getName() + "\"";
  }
}

File: base/src/main/java/de/whitefrog/frogr/persistence/FieldDescriptor.kt
package de.whitefrog.frogr.persistence

import de.whitefrog.frogr.helper.ReflectionUtil
import de.whitefrog.frogr.model.Base
import de.whitefrog.frogr.model.Model
import de.whitefrog.frogr.model.annotation.*
import de.whitefrog.frogr.model.relationship.Relationship

import java.lang.reflect.Field

/**
 * Describes a field on any model and provides easy access to its annotations.
 * Provides methods to check if a field is a collection or a single model.
 */
@Suppress("UNCHECKED_CAST")
class FieldDescriptor<T : Base> internal constructor(private val field: Field) {
  private val annotations: AnnotationDescriptor
  private var baseClass: Class<T>
  private var baseClassName: String
  private val f = field

  /**
   * Indicates that the field is a model type or a collection of models.
   */
  val isModel: Boolean
  /**
   * Indicates that the field is a collection.
   */
  val isCollection: Boolean
  /**
   * Indicates that the field is a relationship type or a collection of relationships.
   */
  val isRelationship: Boolean

  /**
   * Get the fields name.
   */
  val name: String
    get() { return f.name }

  init {
    field.isAccessible = true
    this.isCollection = Collection::class.java.isAssignableFrom(field.type)

    val descriptor = AnnotationDescriptor()
    descriptor.indexed = field.getAnnotation(Indexed::class.java)
    descriptor.notPersistent = field.isAnnotationPresent(NotPersistent::class.java)
    descriptor.relatedTo = field.getAnnotation(RelatedTo::class.java)
    descriptor.unique = field.isAnnotationPresent(Unique::class.java)
    descriptor.fetch = field.getAnnotation(Fetch::class.java)
    descriptor.required = field.isAnnotationPresent(Required::class.java)
    descriptor.nullRemove = field.isAnnotationPresent(NullRemove::class.java)
    descriptor.uuid = field.isAnnotationPresent(Uuid::class.java)
    descriptor.lazy = field.isAnnotationPresent(Lazy::class.java)
    descriptor.relationshipCount = field.getAnnotation(RelationshipCount::class.java)

    this.annotations = descriptor

    if (this.isCollection) {
      this.baseClass = ReflectionUtil.getGenericClass(field) as Class<T>
    } else {
      this.baseClass = field.type as Class<T>
    }
    this.baseClassName = this.baseClass.simpleName

    this.isRelationship = Relationship::class.java.isAssignableFrom(baseClass)
    this.isModel = Model::class.java.isAssignableFrom(baseClass)
  }

  /**
   * Get the annotation descriptor for the described field.
   * @return the annotation descriptor for the described field
   */
  fun annotations(): AnnotationDescriptor {
    return annotations
  }

  /**
   * Get the fields class, or if it's a collection, the generic type.
   * @return the fields class, or if it's a collection, the generic type
   */
  fun baseClass(): Class<T> {
    return baseClass
  }

  /**
   * Get the base class's name.
   * @return the name of the base class
   */
  fun baseClassName(): String {
    return baseClassName
  }

  /**
   * Get the reflected field.
   * @return the reflected field
   */
  fun field(): Field {
    return field
  }

  override fun toString(): String {
    return "Field: \"" + name + "\""
  }
}
